content/docs/arvo/dill/data-types.md

```diff
@@ -0,0 +1,187 @@
++++
+title = "Data Types"
+weight = 4
+template = "doc.html"
++++
+
+Here are the data types used by Dill, as defined in `/sys/lull.hoon`.
+
+## `$blew`
+
+```hoon
+  +$  blew  [p=@ud q=@ud]
+```
+
+Terminal dimension; `p` is columns, `q` is rows. This structure is passed to Dill by the runtime in a [%blew](/docs/arvo/dill/tasks#blew) `task` whenever the dimensions of the terminal changes.
+
+## `$belt`
+
+```hoon
++$  belt
+  $?  bolt
+  $%  [%mod mod=?(%ctl %met %hyp) key=bolt]
+      [%txt p=(list @c)]
+      [%ctl p=@c]
+      [%met p=@c]
+  ==  ==
+```
+
+Terminal input.
+
+A `$belt` is passed to Dill in a [%belt](/docs/arvo/dill/tasks#belt) `task` by the runtime whenever there is input, such as a user typing in the console. This is the old `$belt` and is only used by the runtime, the new [$dill-belt](#dill-belt) is what is used between Dill and userspace.
+
+May either be a [$bolt](#bolt) or one of:
+
+- `%mod` - Modifier (Ctrl, Meta or Hyper) plus [key].
+- `%txt` - A series of characters
+- `%ctl` - Ctrl+[key]
+- `%met` - Meta+[key]
+
+## `$bolt`
+
+```hoon
++$  bolt
+  $@  @c
+  $%  [%aro p=?(%d %l %r %u)]
+      [%bac ~]
+      [%del ~]
+      [%hit r=@ud c=@ud]
+      [%ret ~]
+  ==
+```
+
+Simple input.
+
+Either a single simple character or one of:
+
+- `%aro` - Arrow keys.
+- `%bac` - Backspace key.
+- `%del` - Delete key.
+- `%hit` - Mouse click - `r` is row and `c` is column.
+- `%ret` - Return (Enter) key.
+
+## `$blit`
+
+```hoon
++$  blit
+  $%  [%bel ~]
+      [%clr ~]
+      [%hop p=@ud]
+      [%klr p=stub]
+      [%lin p=(list @c)]
+      [%mor ~]
+      [%sag p=path q=*]
+      [%sav p=path q=@]
+      [%url p=@t]
+  ==
+```
+
+Terminal output.
+
+A `$blit` is given to the runtime by Dill in a [%blit](/docs/arvo/dill/tasks#blit) `task` when it wants to print some text, clear the screen, go _ding_ or what have you.
+
+This is the old `$blit` which is only used between Dill and the runtime, the new [$dill-blit](#dill-blit) is used instead between userspace and Dill.
+
+A `$blit` is one of:
+
+- `%bel` - Ring the terminal bell.
+- `%clr` - Clear the screen.
+- `%hop` - Set cursor position, `p` specifies its position on the prompt line.
+- `%klr` - Set styled line, the `$stub` specifies the text and style.
+- `%lin` - Set current line, `p` contains the text.
+- `%mor` - Newline.
+- `%sag` - Save to jamfile, typically in `/[pier]/.urb/put/`. `p` is `/[optional path]/[filename]/[extension]`. For example, `/foo/bar` will save it in `/[pier]/.urb/put/foo.bar` and `/a/b/c/foo/bar` will save it in `/[pier]/.urb/put/a/b/c/foo.bar`. `q` is the `noun` to `jam` and save in the file.
+- `%sav` - Save to file. Same behaviour as `%sag` except `q` is an `atom` rather than a `noun` and therefore doesn't need to be `jam`med.
```

Comment:
The atom is written to disk as if it were the bytestring in the tail of an `octs`. That is, `%sav`ing the cord `'abcdef'`, whose `@ux` value is `0x6665.6463.6261`, results in a unix file whose hex dump renders as `61 62 63 64 65 66`.